Filter connection-related headers when forwarding HTTP/1.1 responses over H/2

## Summary and context

We are unable to make HTTPS requests in all browsers. 

Here is the request we're doing:

curl -X POST \
  https://endpoint \
  -H 'Content-Type: application/json' \
  -d '{"username":"zdsfasdf","password":"dsfdafsd"}'

Here are the results we are seeing:

 1. Chrome does not work – Fails with:
POST https://endpoint net::ERR_SPDY_PROTOCOL_ERROR
 2. Safari does not work – Fails with:
[Error] Failed to load resource: The operation couldn’t be completed. Protocol error
 3. Curl 7.54.0 does not work – Fails with:
curl: (92) HTTP/2 stream 1 was not closed cleanly: PROTOCOL_ERROR (err 1)
 4. Curl 7.54.0 works when adding the flag --http 1.1
 5. Curl 7.64.1 works
 6. Postman works (after disabling SSL certificate verification)
 7. Firefox works
 
## How to reproduce

Run the RIG on 2.1.1 with this configuration:

1: Here’s our port configuration:
       - name: INBOUND_PORT
          value: "3010"
        - name: INBOUND_HTTPS_PORT
          value: "3011"
        - name: API_HTTP_PORT
          value: "4010"
        - name: API_HTTPS_PORT
          value: "4011"
2: Here’s our https certs config:
        - name: HTTPS_CERTFILE
          value: "/gateway-certs/selfsigned.pem"
        - name: HTTPS_KEYFILE
          value: "/gateway-certs/selfsigned_key.pem"
3: The above certs are the same ones as in the RIG repo

Make a POST request to this server in any of the failing clients above.

> Describe how your clients are set up, whether they connect via SSE or WS, etc.

This is happening in the REST request to authorize, prior to SSE or WS connection.

> A clear and concise description of what you expected to happen.

## Versions (please complete the following information):**

 - Host OS: FROM elixir:1.6-alpine as build
 - Frontend Chrome, Safari, Firefox
 - RIG 2.1.1

-it is a HAPI server with http/1.1
-it does reach the backend service, and the service appears to reply the same way to both HTTP and HTTPS calls
-curl verbose:

```
Note: Unnecessary use of -X or --request, POST is already inferred.
*   Trying IPADDRESS...
* TCP_NODELAY set
* Connected to reactive-gateway-service (IPADDRESS) port 3011 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/certs/ca-certificates.crt
  CApath: none
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-AES256-GCM-SHA384
* ALPN, server accepted to use h2
* Server certificate:
*  subject: O=Phoenix Framework; CN=Self-signed test certificate
*  start date: Dec 13 00:00:00 2018 GMT
*  expire date: Dec 13 00:00:00 2019 GMT
*  issuer: O=Phoenix Framework; CN=Self-signed test certificate
*  SSL certificate verify result: unable to get local issuer certificate (20), continuing anyway.
* Using HTTP2, server supports multi-use
* Connection state changed (HTTP/2 confirmed)
* Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
* Using Stream ID: 1 (easy handle 0x55847c4405e0)
> POST /is/auth HTTP/2
> Host: reactive-gateway-service:3011
> User-Agent: curl/7.61.1
> Accept: */*
> Content-Type: application/json
> Content-Length: 45
>
* Connection state changed (MAX_CONCURRENT_STREAMS == 100)!
* We are completely uploaded and fine
* http2 error: Invalid HTTP header field was received: frame type: 1, stream: 1, name: [connection], value: [keep-alive]
* HTTP/2 stream 0 was not closed cleanly: PROTOCOL_ERROR (err 1)
* stopped the pause stream!
* Connection #0 to host reactive-gateway-service left intact
curl: (92) HTTP/2 stream 0 was not closed cleanly: PROTOCOL_ERROR (err 1)
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -17,6 +17,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Fixed
 
 - Fixed a bug that caused the subscriptions endpoint to return an internal server error when running RIG in a clustered setup. [#194](https://github.com/Accenture/reactive-interaction-gateway/issues/194)
+- Support for forwarding HTTP/1.1 responses over a HTTP/2 connection by dropping connection-related HTTP headers. [#193](https://github.com/Accenture/reactive-interaction-gateway/issues/193)
 
 <!-- ### Deprecated -->
 

--- a/apps/rig_inbound_gateway/lib/rig_inbound_gateway/api_proxy/handler/http.ex
+++ b/apps/rig_inbound_gateway/lib/rig_inbound_gateway/api_proxy/handler/http.ex
@@ -41,6 +41,7 @@ defmodule RigInboundGateway.ApiProxy.Handler.Http do
       req_headers
       |> HttpHeader.put_host_header(url)
       |> HttpHeader.put_forward_header(conn.remote_ip, host_ip)
+      |> drop_connection_related_headers()
 
     result = do_request(method, url, params, req_headers)
 
@@ -213,7 +214,11 @@ defmodule RigInboundGateway.ApiProxy.Handler.Http do
            body: body
          }
        ) do
-    headers = for {k, v} <- headers, do: {String.downcase(k), v}
+    headers =
+      headers
+      |> Enum.map(fn {k, v} -> {String.downcase(k), v} end)
+      |> drop_connection_related_headers()
+
     conn = %{conn | resp_headers: headers}
 
     # only possibility for "response_from" = "http", therefore hardcoded here
@@ -266,5 +271,31 @@ defmodule RigInboundGateway.ApiProxy.Handler.Http do
       response_from,
       "unreachable"
     )
+  end
+
+  # ---
+
+  defp drop_connection_related_headers(headers) do
+    # Connection related headers break HTTP/2
+    # (and it doesn't make sense to forward them anyway).
+    # See https://tools.ietf.org/html/rfc7540#section-8.1.2.2
+
+    headers
+    |> Enum.map(fn {k, v} -> {String.downcase(k), v} end)
+    |> Enum.filter(fn
+      {k, _}
+      when k in [
+             "connection",
+             "keep-alive",
+             "proxy-connection",
+             "transfer-encoding",
+             "upgrade",
+             "http2-settings"
+           ] ->
+        false
+
+      _ ->
+        true
+    end)
   end
 end

